### DISCUSSION TITLE
Creating new permissions dynamically

### DISCUSSION THREAD

**Discussion by kipchirchirian** [codeigniter4/shield#478] [Q&A]
**Qn 1**: How do I create new permissions dynamically without adding it directly in the ` AuthGroups.php` file?
I have tried several ways e.g.

```php
// Example 1
$permissions = [
            'users.verify' => 'Can verify users',
  ];
setting('AuthGroups.permissions', $permissions);

// Example 2
setting('AuthGroups.permissions')[] = $permissions;
```

I have looked into [Bonfire project](https://github.com/lonnieezell/Bonfire2) as answered in discussion #448 and it helped me a lot with other areas however, I found very little info to do with permissions.

**NOTE**: Example 1 works but it overwrites whatever I had added before when I run it the 2nd time.

**Qn 2**: For some reason, I am unable to add permissions that are listed in the `AuthGroups.php` file to the currently logged in user and I get `CodeIgniter\Shield\Authorization\AuthorizationException` when I try it. See the code below:

```php
auth()->user()->addPermission('users.create');
```

This works when I create permissions with _Example 1_ code above and I add the newly created permission to the user.

I guess I am missing something in the 2 problems above or I'm just exhausted. I'll appreciate if anyone helps me point them out.

Edit:
Sorry, had to mention some people to get answers. Cc @lonnieezell @kenjis @MGatner 

Codeigniter 4 version: _4.2.7_
Codeigniter 4 shield version: _dev-develop_

**Comment by grisinformatica** [comment]
Hi, didn't test it but I think you you should use [Settings](https://github.com/codeigniter4/settings) methods `get()` and `set()`.

**Reply by kipchirchirian** [reply]
I think example 1 is a shorthand way of `set()`.

**Reply by grisinformatica** [reply]
Correct, using the helper. My mistake.

**Comment by kipchirchirian** [comment]
For Qn 1: A solution that seems like a hack or a roundabout way of doing it would be:

```php
$newPermissions = [
        'users.verify' => 'Can verify users',
  ];

// 1. Retrieve all permissions
$existingPermissions = setting()->get('AuthGroups.permissions');

// 2. Merge the two arrays containing existing permissions and new permissions
$permissions = array_merge($existingPermissions, $newPermissions);

// 3. Reset the new permissions again
setting()->set('AuthGroups.permissions', $permissions);
```

I'll appreciate any feedback or other solutions on this.
